### PR TITLE
Added feature to exclude entries from being logged (e.g. sensitive data)

### PR DESCRIPTION
**Pull request by benleen**
Hey arb! Had discussed this with @Marsup a while back:

Additional feature to exclude log entries from making it through to being logged.

### CODE REVIEW

**Comment by benleen** [comment]
S`{ log: 'user', ops: '*', request: { include: ['hapi', 'foo'] } }` would be the same as doing `{ log: 'user', ops: '*', request: ['hapi', 'foo'] }`.

You get the extra `exclude` array, which will reject an event if it matches one of the tags, no matter what else is matched in an include.

**Comment by benleen** [comment]
As I added another test for completion I uncovered an existing bug which was duplicating all tags in the mapped object.
All tags were being duplicated in case the tags were an array; `tags = filter` > `tags.concat(filter)`

**Comment by arb** [comment]
This looks nice so far. Can you add some documentation?

**Comment by benleen** [comment]
Not a breaking change. The emphasis is still on the inclusion of tag.
Looking at the tests, I've moved to a different intermediary mapping object in the subscription but all existing `filter` tests remained to output the same results as before.

Will update the documentation!

**Comment by benleen** [comment]
@arb I did notice in your documentation (example) you actually described what I considered to be a bug:
`request: [ 'hapi', 'foo', 'hapi', 'foo' ]` (repetition of tags)

Was this intended?

**Comment by benleen** [comment]
Also, seems I'm wrong about the breaking change. :-)

It IS a breaking if we considering projects using the subscription function directly, which is now returning a different type of object (include/exclude object). Though I expect the amount of people using this is limited?

**Comment by benleen** [comment]
You know what, I'll make that bit backwards compatible as well. ;-)

**Comment by benleen** [comment]
Fully backwards compatible now, including subscription.

**Comment by benleen** [comment]
Readme updated.

**Comment by lloydbenson** [comment]
Thanks for the work @benleen

**Comment by arb** [comment]
Are we just rolling the dice that people aren't using "include" and "exclude" as existing filtering tags? 

As far as repetition of tags... not sure what that was there for of it was a typo. That being said, duplicate tags shouldn't have any impact on the filtering right?

File: README.md [hunk 1]
```diff
@@ -17,7 +17,7 @@ good-squeeze is a collection of small transform streams. The `Squeeze` stream is
 
 Creates a new Squeeze transform stream where:
 
-- `events` an object where each key is a valid good event, and the value is a string or array of strings representing event tags. "\*" indicates no filtering. `null` and `undefined` are assumed to be "\*". Defaults to `{}`
+- `events` an object where each key is a valid good event, and the value is a string, array of strings representing event tags to include or an object defining tags to include and/or exclude. "\*" indicates all tags match. `null` and `undefined` are assumed to be "\*". Defaults to `{}`
```
Comment (arb) [inline README.md]: Can you show an example of the different ways to configure these events... surrounding `include` and `exclude` would help highlight the fact that they are keywords in an object.

File: lib/squeeze.js [hunk 2]
```diff
@@ -65,7 +88,11 @@ class Squeeze extends Stream.Transform {
             let result = false;
             for (let i = 0; i < tags.length; ++i) {
                 const eventTag = tags[i];
-                result = result || subEventTags.indexOf(eventTag) > -1;
+                // As soon as an exclude tag matches, exclude the log entry
+                if (subEventTags.exclude.indexOf(eventTag) > -1) {
+                    return false;
+                }
+                result = result || subEventTags.include.indexOf(eventTag) > -1 || subEventTags.include.length === 0;
```
Comment (arb) [inline lib/squeeze.js]: Are you using `include.length` of `0` to mean don't filter anything and include every event, regardless of tag?

**Comment by benleen** [comment]
To be sure: I was talking about all tags being doubled out in the subscription result. Which seems obsolete. The validation logic will also loop through all `subEventTags` which means it loops twice over the same array of strings.

Will look at dependencies.

**Comment by benleen** [comment]
```Package              Current  Wanted  Latest  Location
code                   2.3.0   2.3.0   3.0.2  good-squeeze
fast-safe-stringify    1.0.9   1.0.9   1.1.0  good-squeeze
hoek                   3.0.4   3.0.4   4.0.2  good-squeeze
lab                   10.9.0  10.9.0  11.0.1  good-squeeze```

**Comment by benleen** [comment]
I see what you mean with people using `include` or `exclude` as filtering tags.

I do wonder if anyone was using the filter or subscription with anything else but a string or array of strings (as the tests were assuming).
Hapi tags are strings & the subscription function is also forcing the input filter into being an array of strings, no matter if the was an object.

That being said though, if we want to be semantically correct we could release it as a breaking change.

File: README.md [hunk 3]
```diff
@@ -17,7 +17,7 @@ good-squeeze is a collection of small transform streams. The `Squeeze` stream is
 
 Creates a new Squeeze transform stream where:
 
-- `events` an object where each key is a valid good event, and the value is a string or array of strings representing event tags. "\*" indicates no filtering. `null` and `undefined` are assumed to be "\*". Defaults to `{}`
+- `events` an object where each key is a valid good event, and the value is a string, array of strings representing event tags to include or an object defining tags to `include` and/or `exclude`. "\*" indicates all tags match. `null` and `undefined` are assumed to be "\*". Defaults to `{}`
```
Comment (benleen) [inline README.md]: @arb About the examples. It didn't seem right to inline them here, so as you noticed I added an example a bit down in this readme next to the other one. Should I add an example or context (similar to my initial comments) on the filter function instead?
Right now the tests are telling the story about how the subscription is interpreted.

File: lib/squeeze.js (hunk 2 above)
Comment (benleen) [inline lib/squeeze.js]: Yes, this is consistent with how the previous 'include' string array was being interpreted:
https://github.com/hapijs/good-squeeze/blob/master/lib/squeeze.js#L59

**Comment by benleen** [comment]
I've updated the readme though I'm not to happy with it either.. Maybe less is more? I'm new to writing this kind of documentation, suggestions are welcome!

File: README.md [hunk 4]
```diff
@@ -10,14 +10,20 @@ Lead Maintainer: [Adam Bretz](https://github.com/arb)
...
-
+    
 ## Methods
 
 ### `Squeeze(events, [options])`
 
 Creates a new Squeeze transform stream where:
 
-- `events` an object where each key is a valid good event, and the value is a string or array of strings representing event tags. "\*" indicates no filtering. `null` and `undefined` are assumed to be "\*". Defaults to `{}`
+- `events` an object where each key is a valid good event, defaults to `{}`. The value which can be one of the following types representing tags to filter
```
Comment (arb) [inline README.md]: Maybe word this something like...

- `events` an object where each key is a valid good event and the value is one of the following:
  - `string` - a tag to include when filtering. `'*'` indicates no filtering. 
  - `array` - an array of tags to include when filtering. `[]` indicates no filtering.
  - `object` - an object with the following values
    - `include` - string or array representing tag(s) to *include* when filtering
    - `exclude` - string or array representing tag(s) to *exclude* when filtering. `exclude` takes precedence over any `include` tags.

**Comment by arb** [comment]
I've decided to just make this a breaking change...

- [ ] Clean up the readme a little more
- [ ] Bump all the outdated deps (and fix the tests)

Then this will be ready to land.

Thanks a lot!

**Comment by benleen** [comment]
Sounds good!

**Comment by benleen** [comment]
Updated dependencies + engine (v6).

Reason tests were updated:
https://github.com/hapijs/code/commit/710aaf0868108251ab248a6f1d7208004d2544ea#diff-88333ba11114e014e87198d04680144dR95

**Comment by benleen** [comment]
Should be good now. Apologies for the force pushes (just wanted to keep the commits a bit cleaner).

**Pull request merged**